src/librustc_trans/cabi_mips64.rs

```diff
@@ -8,50 +8,153 @@
+                        layout::Abi::Scalar(ref scalar) => match scalar.value {
+                            layout::F32 => Some(Reg::f32()),
+                            layout::F64 => Some(Reg::f64()),
+                            _ => None
+                        },
+                        _ => None
+                    }
+                }).collect();
+
+            if ret.layout.fields.count() == float_regs.len() {
+                if float_regs.len() == 1 {
+                    ret.cast_to(Uniform { unit: float_regs[0], total: float_regs[0].size });
+                    return;
+                } else if float_regs.len() == 2 {
+                    ret.cast_to(CastTarget::Pair(float_regs[0], float_regs[1]));
+                    return;
+                }
+            }
+        }
+
+        // Cast to a uniform int structure
+        ret.cast_to(Uniform {
+            unit: bits_to_int_reg(bits),
+            total: size
+        });
     } else {
         ret.make_indirect();
-        *offset += cx.tcx.data_layout.pointer_size;
     }
 }
 
-fn classify_arg_ty(cx: &CodegenCx, arg: &mut ArgType, offset: &mut Size) {
+fn classify_arg_ty<'a, 'tcx>(cx: &CodegenCx<'a, 'tcx>, arg: &mut ArgType<'tcx>) {
     let dl = &cx.tcx.data_layout;
     let size = arg.layout.size;
-    let align = arg.layout.align.max(dl.i32_align).min(dl.i64_align);
 
     if arg.layout.is_aggregate() {
-        arg.cast_to(Uniform {
-            unit: Reg::i64(),
+        let mut chunks = Vec::new();
+        match arg.layout.fields {
+            layout::FieldPlacement::Array { .. } => {
+                // Arrays are passed indirectly
+                arg.make_indirect();
+                return;
+            }
+            layout::FieldPlacement::Union(_) => {
+                // Unions and are always treated as a series of 64-bit integer chunks
+            },
+            layout::FieldPlacement::Arbitrary { .. } => {
+                // Structures are split up into a series of 64-bit integer chunks, but any aligned
+                // doubles not part of another aggregate are passed as floats.
+                let mut last_offset = Size::from_bytes(0);
+
+                for i in 0..arg.layout.fields.count() {
+                    let field = arg.layout.field(cx, i);
+                    let offset = arg.layout.fields.offset(i);
+
+                    // We only care about aligned doubles
+                    if let layout::Abi::Scalar(ref scalar) = field.abi {
+                        if let layout::F64 = scalar.value {
+                            if offset.is_abi_aligned(dl.f64_align) {
+                                // Insert enough integers to cover [last_offset, offset)
+                                assert!(last_offset.is_abi_aligned(dl.f64_align));
+                                for _ in 0..((offset - last_offset).bits() / 64) {
+                                    chunks.push(RegKind::Integer);
```

Comment:
You can keep an index into the final array to avoid allocating a `Vec`.

Reply:
Fixed.